File: src/medium/Q0099.py
"""
99. Recover Binary Search Tree
https://leetcode.com/problems/recover-binary-search-tree/
"""
from __future__ import annotations
from dataclasses import dataclass


@dataclass
class TreeNode:
    val: int
    left: TreeNode | None
    right: TreeNode | None


# Time: O(n), Space: O(n), where n is the number of nodes in the tree.
def recover_tree(root: TreeNode | None) -> None:
    nodes = []

    def dfs(node: TreeNode | None) -> None:
        if node is None:
            return
        dfs(node.left)
        nodes.append(node)
        dfs(node.right)

    dfs(root)
    sorted_vals = sorted([node.val for node in nodes])
    for i in range(len(nodes)):
        nodes[i].val = sorted_vals[i]

File: src/medium/Q0099.cpp
// 99. Recover Binary Search Tree
// https://leetcode.com/problems/recover-binary-search-tree/

#include <functional>
#include <vector>

using namespace std;

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
};

// Time: O(n), Space: O(n), where n is the number of nodes in the tree.
void recover_tree(TreeNode *root) {
  auto nodes = vector<TreeNode *>();

  function<void(TreeNode *)> dfs = [&dfs, &nodes](TreeNode *node) {
    if (node == nullptr) {
      return;
    }
    dfs(node->left);
    nodes.push_back(node);
    dfs(node->right);
  };

  dfs(root);

  auto sorted_vals = vector<int>();
  sorted_vals.reserve(nodes.size());
  for (const auto &node : nodes) {
    sorted_vals.push_back(node->val);
  }
  sort(sorted_vals.begin(), sorted_vals.end());

  for (int i = 0; i < nodes.size(); ++i) {
    nodes[i]->val = sorted_vals[i];
  }
}
